_posts/2020-08-05-using-open-policy-agent-with-strimzi-and-apache-kafka.md

```diff
@@ -0,0 +1,451 @@
+    - CN=henri
+    - anwar
+    - CN=wesley
+```
+
+### Examples policies
+
+Explaining the basics of the Rego language and Open Policy Agent policies goes well beyond the scope of this blog post.
+Both are nicely explained in the [OPA documentation](https://www.openpolicyagent.org/docs/latest/).
+But I will show you some examples of how you can use OPA with Apache Kafka to demonstrate its strength and versatility.
+The complete policy files from these examples including the instructions on how to use them can be found on my [GitHub](https://github.com/scholzj/demo-opa-kafka-authorization).
+
+Additionally, the OPA documentation has a simple demo for [Kafka authorization](https://www.openpolicyagent.org/docs/latest/kafka-authorization/) as well.
+
+### Group based authorization
+
+One of the features Strimzi users are asking for is group based authorization.
+This is currently not supported in Strimzi using the `simple` authorization. 
+But this example shows how it can be implemented using Open Policy Agent.
+
+We will use 3 different groups: consumers, producers and admins.
+For each of these groups, we will need to define the set of operations they will be allowed to use.
+We will also add some helper rules to help us evaluate whether the operation we are authorizing belongs into one of these groups or not:
+
+```rego
+consumer_operations = {
+                        "Topic": ["Read", "Describe"], 
+                        "Group": ["Read", "Describe"]
+                    }
+
+producer_operations = {
+                        "Topic": ["Write", "Describe"]
+                    }
+
+admin_operations = {
+                    "Topic": ["Read", "Write", "Create", "Delete", "Alter", "Describe", "ClusterAction", "DescribeConfigs", "AlterConfigs"], 
+                    "Group": ["Read", "Delete", "Describe"],
+                    "Cluster": ["Create", "Alter", "Describe", "ClusterAction", "DescribeConfigs", "AlterConfigs"]
+                    }
+
+is_consumer_operation {
+    consumer_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+
+is_producer_operation {
+    producer_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+
+is_admin_operation {
+    admin_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+```
+
+We also need to define the actual groups and their members:
+
+```rego
+consumer_group = ["tom", "tyrone", "matt", "pepe", "douglas"]
+producer_group = ["jack", "conor", "keinan", "john"]
+admin_group = ["dean", "christian"]
+
+is_consumer_group {
+    consumer_group[_] == principal.name
+}
+
+is_producer_group {
+    producer_group[_] == principal.name
+}
+
+is_admin_group {
+    admin_group[_] == principal.name
+}
+```
+
+The group definition is accompanied by some helper rules.
+Each group is defined as an array of its members.
+So we have the groups and their members hardcoded into the policy and to change them, you will need to change the policy.
```

Comment:
does the OPA server loads the policy at runtime if you change it?

Reply:
You can do it using the REST API for example. But how to do it and hot to configure OPA is really out of scope for the blog post. It is not used in the example in my GitHub repo.